src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -897,6 +897,122 @@ public void SendRecvIovMax_Success()
+            using (var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            {
+                socket.BindToAnonymousPort(IPAddress.Loopback);
+                // Use our own address as destination.
+                socket.Connect(socket.LocalEndPoint);
+
+                var sendBuffer = new byte[segmentCount];
+                var sendSegments = new List<ArraySegment<byte>>();
+                for (int i = 0; i < segmentCount; i++)
+                {
+                    sendBuffer[i] = (byte)i;
+                    sendSegments.Add(new ArraySegment<byte>(sendBuffer, i, 1));
+                }
+
+                SocketError error;
+                // send data as segmentCount (> IOV_MAX) 1-byte segments.
+                int bytesSent = socket.Send(sendSegments, SocketFlags.None, out error);
+                if (error == SocketError.Success)
+                {
+                    // platform sent message with > IOV_MAX segments
+                    Assert.Equal(segmentCount, bytesSent);
+                }
+                else
+                {
+                    // platform returns EMSGSIZE
+                    Assert.Equal(SocketError.MessageSize, error);
+                    Assert.Equal(0, bytesSent);
+                }
+            }
+        }
+
+        [Fact]
+        public async Task ReceiveIovMaxUdp_SuccessOrMessageSize()
+        {
+            // receiving more than IOV_MAX segments causes EMSGSIZE on some platforms.
+            // We handle this for stream sockets by truncating.
+            // This test verifies we are not truncating non-stream sockets.
+
+            // Use more than IOV_MAX (1024 on Linux & macOS) segments
+            // and less than Ethernet MTU.
+            const int segmentCount = 1200;
+            var receiver = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+            receiver.BindToAnonymousPort(IPAddress.Loopback);
+            Task receiveTask = Task.Run(() =>
+            {
+                using (receiver)
+                {
+                    var receiveBuffer = new byte[segmentCount];
+                    var receiveSegments = new List<ArraySegment<byte>>();
+                    for (int i = 0; i < segmentCount; i++)
+                    {
+                        receiveSegments.Add(new ArraySegment<byte>(receiveBuffer, i, 1));
+                    }
+                    // receive data as segmentCount (> IOV_MAX) 1-byte segments.
+                    EventWaitHandle completed = new ManualResetEvent(false);
+                    var socketEvent = new SocketAsyncEventArgs();
+                    socketEvent.UserToken = completed;
+                    socketEvent.BufferList = receiveSegments;
+                    socketEvent.Completed += OnCompleted;
+                    if (receiver.ReceiveAsync(socketEvent))
+                    {
+                        completed.WaitOne();
```

Comment:
Rather than blocking the thread, you can use a TaskCompletionSource instead of a ManualResetEvent, await completed.Task instead of completed.WaitOne(), and completed.SetResult instead of completed.Set.